File: app/ui/streaming_types.py
"""
流式消息类型定义模块
基于LangGraph和Vercel AI SDK设计的统一流式消息架构
"""

from enum import Enum
from typing import Dict, Any, Optional, List, Union
from datetime import datetime
from pydantic import BaseModel, Field
import uuid


class StreamMessageType(str, Enum):
    """流式消息类型枚举"""
    # 节点状态相关
    NODE_START = "node_start"           # 节点开始执行
    NODE_COMPLETE = "node_complete"     # 节点执行完成
    NODE_ERROR = "node_error"           # 节点执行错误
    
    # 路由相关
    ROUTE_DECISION = "route_decision"   # 路由决策
    ROUTE_CHANGE = "route_change"       # 路由变化
    
    # LLM相关
    LLM_TOKEN = "llm_token"            # LLM token流
    LLM_COMPLETE = "llm_complete"      # LLM回答完成
    
    # 工具执行相关
    TOOL_START = "tool_start"          # 工具开始执行
    TOOL_PROGRESS = "tool_progress"    # 工具执行进度
    TOOL_COMPLETE = "tool_complete"    # 工具执行完成
    TOOL_ERROR = "tool_error"          # 工具执行错误
    
    # 文件相关
    FILE_GENERATED = "file_generated"  # 文件生成
    FILE_UPLOADED = "file_uploaded"    # 文件上传
    
    # Agent思考相关
    AGENT_THINKING = "agent_thinking"  # Agent思考过程
    AGENT_ANALYSIS = "agent_analysis" # Agent分析结果
    
    # 系统状态相关
    SESSION_START = "session_start"    # 会话开始
    SESSION_END = "session_end"        # 会话结束
    ERROR = "error"                    # 通用错误
    INFO = "info"                      # 通用信息
    DEBUG = "debug"                    # 调试信息


class MessagePriority(str, Enum):
    """消息优先级"""
    LOW = "low"
    NORMAL = "normal" 
    HIGH = "high"
    URGENT = "urgent"


class BaseStreamMessage(BaseModel):
    """流式消息基类"""
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    type: StreamMessageType
    timestamp: datetime = Field(default_factory=datetime.now)
    session_id: Optional[str] = None
    source: Optional[str] = None  # 来源节点/智能体名称
    priority: MessagePriority = MessagePriority.NORMAL
    metadata: Dict[str, Any] = Field(default_factory=dict)


class NodeStatusMessage(BaseStreamMessage):
    """节点状态消息"""
    node_name: str
    node_type: Optional[str] = None  # supervisor, agent, tool等
    status: str  # started, completed, error
    progress: Optional[float] = None  # 0.0-1.0
    details: Optional[str] = None
    duration: Optional[float] = None  # 执行耗时（秒）


class RouterMessage(BaseStreamMessage):
    """路由消息"""
    from_node: Optional[str] = None
    to_node: str
    reason: Optional[str] = None  # 路由原因
    available_routes: Optional[List[str]] = None


class LLMTokenMessage(BaseStreamMessage):
    """LLM Token消息"""
    content: str  # token内容
    is_complete: bool = False  # 是否完整回答完成
    llm_model: Optional[str] = None  # 重命名避免model_命名空间冲突
    token_count: Optional[int] = None


class ToolExecutionMessage(BaseStreamMessage):
    """工具执行消息"""
    tool_name: str
    action: str  # start, progress, complete, error
    input_params: Optional[Dict[str, Any]] = None
    output: Optional[Any] = None
    progress: Optional[float] = None  # 0.0-1.0
    error_message: Optional[str] = None
    execution_time: Optional[float] = None


class FileGeneratedMessage(BaseStreamMessage):
    """文件生成消息"""
    file_id: str
    file_name: str
    file_type: str  # image, document, data等
    file_path: str
    file_size: Optional[int] = None
    category: str = "generated"  # upload, generated, temp
    folder_path: Optional[str] = None  # 在文件树中的路径
    preview_url: Optional[str] = None  # 预览URL
    download_url: Optional[str] = None  # 下载URL


class AgentThinkingMessage(BaseStreamMessage):
    """Agent思考消息"""
    agent_name: str
    thinking_type: str  # analysis, planning, reasoning等
    content: str
    confidence: Optional[float] = None  # 0.0-1.0
    step_number: Optional[int] = None


class ErrorMessage(BaseStreamMessage):
    """错误消息"""
    error_code: Optional[str] = None
    error_message: str
    error_details: Optional[Dict[str, Any]] = None
    recoverable: bool = True
    suggested_action: Optional[str] = None


class InfoMessage(BaseStreamMessage):
    """信息消息"""
    title: Optional[str] = None
    content: str
    level: str = "info"  # info, warning, success
    actionable: bool = False
    action_label: Optional[str] = None
    action_data: Optional[Dict[str, Any]] = None


# 消息类型映射
MESSAGE_TYPE_MAP = {
    StreamMessageType.NODE_START: NodeStatusMessage,
    StreamMessageType.NODE_COMPLETE: NodeStatusMessage,
    StreamMessageType.NODE_ERROR: NodeStatusMessage,
    StreamMessageType.ROUTE_DECISION: RouterMessage,
    StreamMessageType.ROUTE_CHANGE: RouterMessage,
    StreamMessageType.LLM_TOKEN: LLMTokenMessage,
    StreamMessageType.LLM_COMPLETE: LLMTokenMessage,
    StreamMessageType.TOOL_START: ToolExecutionMessage,
    StreamMessageType.TOOL_PROGRESS: ToolExecutionMessage,
    StreamMessageType.TOOL_COMPLETE: ToolExecutionMessage,
    StreamMessageType.TOOL_ERROR: ToolExecutionMessage,
    StreamMessageType.FILE_GENERATED: FileGeneratedMessage,
    StreamMessageType.FILE_UPLOADED: FileGeneratedMessage,
    StreamMessageType.AGENT_THINKING: AgentThinkingMessage,
    StreamMessageType.ERROR: ErrorMessage,
    StreamMessageType.INFO: InfoMessage,
}


def create_message(
    message_type: StreamMessageType,
    session_id: Optional[str] = None,
    source: Optional[str] = None,
    **kwargs
) -> BaseStreamMessage:
    """
    工厂函数：创建指定类型的流式消息
    
    Args:
        message_type: 消息类型
        session_id: 会话ID
        source: 消息来源
        **kwargs: 其他参数
        
    Returns:
        对应类型的消息实例
    """
    message_class = MESSAGE_TYPE_MAP.get(message_type, InfoMessage)
    
    return message_class(
        type=message_type,
        session_id=session_id,
        source=source,
        **kwargs
    )


def serialize_message(message: BaseStreamMessage) -> Dict[str, Any]:
    """
    序列化消息为字典格式，用于JSON传输
    
    Args:
        message: 消息实例
        
    Returns:
        序列化后的字典
    """
    data = message.model_dump()
    data["timestamp"] = message.timestamp.isoformat()
    return data


def deserialize_message(data: Dict[str, Any]) -> BaseStreamMessage:
    """
    从字典反序列化消息
    
    Args:
        data: 序列化的消息数据
        
    Returns:
        消息实例
    """
    message_type = StreamMessageType(data["type"])
    message_class = MESSAGE_TYPE_MAP.get(message_type, InfoMessage)
    
    # 转换时间戳
    if "timestamp" in data and isinstance(data["timestamp"], str):
        data["timestamp"] = datetime.fromisoformat(data["timestamp"])
    
    return message_class(**data)


# 类型别名，用于兼容性
StreamMessage = BaseStreamMessage  # 主要的流式消息类型
LLMCompleteMessage = LLMTokenMessage  # LLM完成消息，复用Token消息类
SystemMessage = InfoMessage  # 系统消息，复用Info消息类

# 导出列表
__all__ = [
    # 枚举
    "StreamMessageType",
    "MessagePriority",
    
    # 基础类
    "BaseStreamMessage",
    "StreamMessage",  # 类型别名
    
    # 具体消息类
    "NodeStatusMessage",
    "RouterMessage", 
    "LLMTokenMessage",
    "LLMCompleteMessage",  # 类型别名
    "ToolExecutionMessage",
    "FileGeneratedMessage",
    "AgentThinkingMessage",
    "ErrorMessage",
    "InfoMessage",
    "SystemMessage",  # 类型别名
    
    # 工具函数
    "create_message",
    "serialize_message",
    "deserialize_message",
    "MESSAGE_TYPE_MAP"
] 

File: app/ui/petro_agent/lib/streaming-types.ts
// 前端流式消息类型定义
// 与后端 app/ui/streaming_types.py 保持一致

export enum StreamMessageType {
  // 节点状态消息
  NODE_START = "node_start",
  NODE_END = "node_end", 
  NODE_ERROR = "node_error",
  
  // 路由决策消息
  ROUTER_DECISION = "router_decision",
  ROUTER_PATH = "router_path",
  
  // LLM相关消息
  LLM_TOKEN = "llm_token",
  LLM_START = "llm_start",
  LLM_END = "llm_end",
  
  // 工具/任务执行消息
  TOOL_START = "tool_start",
  TOOL_PROGRESS = "tool_progress", 
  TOOL_RESULT = "tool_result",
  TOOL_ERROR = "tool_error",
  
  // 文件操作消息
  FILE_GENERATED = "file_generated",
  FILE_UPLOADED = "file_uploaded",
  FILE_PROCESSED = "file_processed",
  
  // Agent思考过程
  AGENT_THINKING = "agent_thinking",
  AGENT_PLANNING = "agent_planning",
  AGENT_DECISION = "agent_decision",
  
  // 系统消息
  SYSTEM_INFO = "system_info",
  SYSTEM_WARNING = "system_warning", 
  SYSTEM_ERROR = "system_error",
  
  // 会话状态
  SESSION_START = "session_start",
  SESSION_END = "session_end",
  
  // 检查点
  CHECKPOINT_SAVED = "checkpoint_saved",
  CHECKPOINT_RESTORED = "checkpoint_restored"
}

// 基础流式消息接口
export interface BaseStreamMessage {
  id: string;
  type: StreamMessageType;
  timestamp: string;
  session_id?: string;
  source?: string;
  priority?: 'low' | 'normal' | 'high' | 'critical';
  metadata?: Record<string, any>;
}

// 节点状态消息
export interface NodeStatusMessage extends BaseStreamMessage {
  type: StreamMessageType.NODE_START | StreamMessageType.NODE_END | StreamMessageType.NODE_ERROR;
  node_id: string;
  node_name: string;
  agent_name?: string;
  details?: string;
  error_info?: string;
}

// 路由决策消息
export interface RouterMessage extends BaseStreamMessage {
  type: StreamMessageType.ROUTER_DECISION | StreamMessageType.ROUTER_PATH;
  decision: string;
  available_paths?: string[];
  selected_path?: string;
  confidence?: number;
  reasoning?: string;
}

// LLM Token消息
export interface LLMTokenMessage extends BaseStreamMessage {
  type: StreamMessageType.LLM_TOKEN | StreamMessageType.LLM_START | StreamMessageType.LLM_END;
  content: string;
  model_name?: string;
  token_count?: number;
  is_final?: boolean;
}

// 工具执行消息  
export interface ToolExecutionMessage extends BaseStreamMessage {
  type: StreamMessageType.TOOL_START | StreamMessageType.TOOL_PROGRESS | StreamMessageType.TOOL_RESULT | StreamMessageType.TOOL_ERROR;
  tool_name: string;
  progress?: number;
  status?: string;
  result?: any;
  error_message?: string;
  execution_time?: number;
}

// 文件生成消息
export interface FileGeneratedMessage extends BaseStreamMessage {
  type: StreamMessageType.FILE_GENERATED | StreamMessageType.FILE_UPLOADED | StreamMessageType.FILE_PROCESSED;
  file_id: string;
  file_name: string;
  file_path: string;
  file_type: string;
  file_size?: number;
  category?: string;
  description?: string;
  thumbnail_url?: string;
}

// Agent思考消息
export interface AgentThinkingMessage extends BaseStreamMessage {
  type: StreamMessageType.AGENT_THINKING | StreamMessageType.AGENT_PLANNING | StreamMessageType.AGENT_DECISION;
  agent_name: string;
  content: string;
  reasoning_step?: number;
  total_steps?: number;
  decision_options?: string[];
  selected_option?: string;
}

// 系统消息
export interface SystemMessage extends BaseStreamMessage {
  type: StreamMessageType.SYSTEM_INFO | StreamMessageType.SYSTEM_WARNING | StreamMessageType.SYSTEM_ERROR;
  message: string;
  details?: string;
  action_required?: boolean;
  suggested_actions?: string[];
}

// 会话状态消息
export interface SessionMessage extends BaseStreamMessage {
  type: StreamMessageType.SESSION_START | StreamMessageType.SESSION_END;
  session_info?: Record<string, any>;
}

// 检查点消息
export interface CheckpointMessage extends BaseStreamMessage {
  type: StreamMessageType.CHECKPOINT_SAVED | StreamMessageType.CHECKPOINT_RESTORED;
  checkpoint_id: string;
  checkpoint_data?: any;
}

// 所有消息类型的联合类型
export type StreamMessage = 
  | NodeStatusMessage
  | RouterMessage  
  | LLMTokenMessage
  | ToolExecutionMessage
  | FileGeneratedMessage
  | AgentThinkingMessage
  | SystemMessage
  | SessionMessage
  | CheckpointMessage;

// 消息解析工厂函数
export function parseStreamMessage(data: any): StreamMessage | null {
  try {
    if (!data || !data.type) {
      return null;
    }

    const baseFields = {
      id: data.id || `msg-${Date.now()}-${Math.random()}`,
      timestamp: data.timestamp || new Date().toISOString(),
      session_id: data.session_id,
      source: data.source,
      priority: data.priority || 'normal',
      metadata: data.metadata || data.data || {}
    };

    switch (data.type) {
      case 'node_start':
      case 'node_complete':
      case 'node_error':
        // 适配后端实际发送的格式
        return {
          ...baseFields,
          type: data.type === 'node_complete' ? StreamMessageType.NODE_END : 
                data.type === 'node_start' ? StreamMessageType.NODE_START : StreamMessageType.NODE_ERROR,
          node_id: data.node_name || data.node_id || 'unknown',
          node_name: data.node_name || 'unknown',
          agent_name: data.agent_name || data.source,
          details: data.details || data.content,
          error_info: data.error_message || (data.type === 'node_error' ? data.details : undefined)
        } as NodeStatusMessage;

      case 'route_decision':
      case StreamMessageType.ROUTER_DECISION:
      case StreamMessageType.ROUTER_PATH:
        return {
          ...baseFields,
          type: StreamMessageType.ROUTER_DECISION,
          decision: data.reason || data.decision || 'unknown',
          available_paths: data.available_routes || data.available_paths,
          selected_path: data.to_node || data.selected_path,
          confidence: data.confidence,
          reasoning: data.reason || data.reasoning
        } as RouterMessage;

      case 'llm_token':
      case StreamMessageType.LLM_TOKEN:
      case StreamMessageType.LLM_START:
      case StreamMessageType.LLM_END:
        return {
          ...baseFields,
          type: StreamMessageType.LLM_TOKEN,
          content: data.content || data.token || '',
          model_name: data.llm_model || data.model_name,
          token_count: data.token_count,
          is_final: data.is_complete || data.is_final || false
        } as LLMTokenMessage;

      case 'tool_start':
      case 'tool_progress':
      case 'tool_complete':
      case 'tool_error':
      case StreamMessageType.TOOL_START:
      case StreamMessageType.TOOL_PROGRESS:
      case StreamMessageType.TOOL_RESULT:
      case StreamMessageType.TOOL_ERROR:
        const toolType = data.type === 'tool_complete' ? StreamMessageType.TOOL_RESULT : 
                        data.type === 'tool_start' ? StreamMessageType.TOOL_START :
                        data.type === 'tool_progress' ? StreamMessageType.TOOL_PROGRESS : StreamMessageType.TOOL_ERROR;
        return {
          ...baseFields,
          type: toolType,
          tool_name: data.tool_name || 'unknown',
          progress: data.progress,
          status: data.status || data.action,
          result: data.output || data.result,
          error_message: data.error_message,
          execution_time: data.execution_time || data.duration
        } as ToolExecutionMessage;

      case 'file_generated':
      case StreamMessageType.FILE_GENERATED:
      case StreamMessageType.FILE_UPLOADED:
      case StreamMessageType.FILE_PROCESSED:
        return {
          ...baseFields,
          type: StreamMessageType.FILE_GENERATED,
          file_id: data.file_id || `file-${Date.now()}`,
          file_name: data.file_name || 'unknown',
          file_path: data.file_path || '',
          file_type: data.file_type || 'unknown',
          file_size: data.file_size,
          category: data.category || 'generated',
          description: data.description,
          thumbnail_url: data.thumbnail_url || data.preview_url
        } as FileGeneratedMessage;

      case 'agent_thinking':
      case StreamMessageType.AGENT_THINKING:
      case StreamMessageType.AGENT_PLANNING:
      case StreamMessageType.AGENT_DECISION:
        // 适配后端实际发送的格式
        let content = data.content;
        let agent_name = data.agent_name;
        
        // 处理后端发送的简化格式: {'agent_thinking': '内容'}
        if (typeof data.agent_thinking === 'string') {
          content = data.agent_thinking;
          agent_name = data.source || 'Agent';
        }
        
        // 提取思考内容中的智能体名称
        if (content && content.includes('正在思考:')) {
          const match = content.match(/🤔\s*(\w+)\s*正在思考:\s*(.+)/);
          if (match) {
            agent_name = match[1];
            content = match[2];
          }
        }
        
        return {
          ...baseFields,
          type: StreamMessageType.AGENT_THINKING,
          agent_name: agent_name || data.source || 'Agent',
          content: content || '',
          reasoning_step: data.reasoning_step || data.step_number,
          total_steps: data.total_steps || 5,
          decision_options: data.decision_options,
          selected_option: data.selected_option
        } as AgentThinkingMessage;

      case StreamMessageType.SYSTEM_INFO:
      case StreamMessageType.SYSTEM_WARNING:
      case StreamMessageType.SYSTEM_ERROR:
        return {
          ...baseFields,
          type: data.type,
          message: data.message || data.content,
          details: data.details,
          action_required: data.action_required,
          suggested_actions: data.suggested_actions
        } as SystemMessage;

      case StreamMessageType.SESSION_START:
      case StreamMessageType.SESSION_END:
        return {
          ...baseFields,
          type: data.type,
          session_info: data.session_info
        } as SessionMessage;

      case StreamMessageType.CHECKPOINT_SAVED:
      case StreamMessageType.CHECKPOINT_RESTORED:
        return {
          ...baseFields,
          type: data.type,
          checkpoint_id: data.checkpoint_id,
          checkpoint_data: data.checkpoint_data
        } as CheckpointMessage;

      default:
        console.warn('未知的流式消息类型:', data.type, data);
        return null;
    }
  } catch (error) {
    console.error('解析流式消息失败:', error, data);
    return null;
  }
}

// 消息类型判断工具函数
export function isNodeMessage(msg: StreamMessage): msg is NodeStatusMessage {
  return [StreamMessageType.NODE_START, StreamMessageType.NODE_END, StreamMessageType.NODE_ERROR].includes(msg.type);
}

export function isRouterMessage(msg: StreamMessage): msg is RouterMessage {
  return [StreamMessageType.ROUTER_DECISION, StreamMessageType.ROUTER_PATH].includes(msg.type);
}

export function isLLMMessage(msg: StreamMessage): msg is LLMTokenMessage {
  return [StreamMessageType.LLM_TOKEN, StreamMessageType.LLM_START, StreamMessageType.LLM_END].includes(msg.type);
}

export function isToolMessage(msg: StreamMessage): msg is ToolExecutionMessage {
  return [StreamMessageType.TOOL_START, StreamMessageType.TOOL_PROGRESS, StreamMessageType.TOOL_RESULT, StreamMessageType.TOOL_ERROR].includes(msg.type);
}

export function isFileMessage(msg: StreamMessage): msg is FileGeneratedMessage {
  return [StreamMessageType.FILE_GENERATED, StreamMessageType.FILE_UPLOADED, StreamMessageType.FILE_PROCESSED].includes(msg.type);
}

export function isAgentMessage(msg: StreamMessage): msg is AgentThinkingMessage {
  return [StreamMessageType.AGENT_THINKING, StreamMessageType.AGENT_PLANNING, StreamMessageType.AGENT_DECISION].includes(msg.type);
}

export function isSystemMessage(msg: StreamMessage): msg is SystemMessage {
  return [StreamMessageType.SYSTEM_INFO, StreamMessageType.SYSTEM_WARNING, StreamMessageType.SYSTEM_ERROR].includes(msg.type);
} 